### PR TITLE
[Serverless] Extract trace context from SQS `AWSTraceHeader` Attribute.

### PR DESCRIPTION
**Pull request by purple4reina**
### What does this PR do?

This pull request makes it possible for universal instrumentation runtimes to accept trace propagation context from an SQS message that was sent using the java tracer.

Instead of adding a `Record.MessageAttribute` like the other tracers, the java tracer uses the `Record.Attribute["AWSTraceHeader"]` to propagate context. This is the xray header added by aws. The java tracer replaces the contents of this attribute with the trace and parent ids.

For more information, see how the java tracer extracts [context from xray headers](https://github.com/DataDog/dd-trace-java/blob/822ffc03d9317dfa4cbb6797038d600ba3659f83/dd-trace-core/src/main/java/datadog/trace/core/propagation/XRayHttpCodec.java#L191-L236) and the [RFC doc](https://docs.google.com/document/d/1zeOv8TTcFUQ7kHFBKwHuoHLxCIqFWr1F95AZUFowmj8/edit) detailing the decision.

### Motivation

Many customers have requested support for trace propagation through SQS for Java.

### Additional Notes

🏅 🏆 I suggest you review this PR one commit at a time because so much of this PR is moving methods from one file to another.

### Possible Drawbacks / Trade-offs

This trace propagation only works from universal instrumented functions: Java, Dotnet, Golang and Ruby. Further work will need to be done to add this support to the Python and Node lambda tracing libraries.

This current implementation only supports 64bit trace ids.

### Describe how to test/QA your changes

Create two java lambdas, one that puts a message onto an SQS queue and the other that reads it. See a connected trace in the UI.

Here's the results from when I did this in my own testing:

Screenshot 2023-07-21 at 2 21 25 PM

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/serverless/trace/inferredspan/propagation.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+	rootPrefix    = "Root="
+	parentPrefix  = "Parent="
+	rootPadding   = len(rootPrefix + "1-00000000-00000000")
+	parentPadding = len(parentPrefix)
+)
+
+var rootRegex = regexp.MustCompile("Root=1-[0-9a-fA-F]{8}-00000000[0-9a-fA-F]{16}")
+
+type rawTraceContext struct {
+	TraceID  string `json:"x-datadog-trace-id"`
```
Comment (astuyve) [inline pkg/serverless/trace/inferredspan/propagation.go:27]: We need sample priority too, right?

Comment (purple4reina) [reply pkg/serverless/trace/inferredspan/propagation.go:27]: Yes agreed. We didn't already support it for the existing SQS trace propagation via `x-datadog-*` headers, so I opted not to include it here. I'll file a separate ticket and we can add sample priority support for both header times in a single PR.

File: pkg/serverless/trace/inferredspan/propagation.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+func extractTraceContextfromAWSTraceHeader(value string) *rawTraceContext {
+	if !rootRegex.MatchString(value) {
+		return nil
+	}
+
+	var startPart int
+	traceData := &rawTraceContext{base: 16}
+	length := len(value)
+	for startPart < length {
```
Comment (astuyve) [inline pkg/serverless/trace/inferredspan/propagation.go:121]: Did we re-implement this algorithm from the java side? Is this canonical, etc?

Comment (purple4reina) [reply pkg/serverless/trace/inferredspan/propagation.go:121]: It's copied from where the java tracer extracts these headers: https://github.com/DataDog/dd-trace-java/blob/822ffc03d9317dfa4cbb6797038d600ba3659f83/dd-trace-core/src/main/java/datadog/trace/core/propagation/XRayHttpCodec.java#L191-L237

File: pkg/serverless/trace/inferredspan/propagation.go (hunk 1 above)
Comment (astuyve) [reply pkg/serverless/trace/inferredspan/propagation.go:27]: Okay, thanks!

File: pkg/serverless/trace/inferredspan/propagation.go (hunk 2 above)
Comment (astuyve) [reply pkg/serverless/trace/inferredspan/propagation.go:121]: Roger that

Comment (purple4reina) [reply pkg/serverless/trace/inferredspan/propagation.go:121]: I'm not sure what the best way ensure the insertion in the java tracer matches our extraction in the extension. I'm thinking adding an SQS tracing test app to self monitoring service is a good start though.

**Review by astuyve — APPROVED** [review]

**Pull request merged**